broadinstitute/single_cell_portal_core
# Adding internal Terra workspaces for studies (SCP-4949)

#### BACKGROUND
The first step of future scaling work related to pre-rendering of visualization data was to determine where such data would be held.  Currently, all SCP studies have an associated Terra workspace where all uploaded files are stored, along with any internal SCP assets, such as ingest log files and some existing rendered data in the form of differential expression outputs.  After evaluating several different options, the decision was made to add a second Terra workspace to all SCP studies where all internal assets will be stored.  Read access can then be granted to users for these internal workspace (write access for admins) to enable future visualization requests. 

#### CHANGES
Now, all studies will have a second Terra workspace created after the primary user-facing one has been successfully provisioned.  These workspaces will have deterministic names of `{accession}-{environment}-internal` (in the `test` environment, a 5-character slug is appended to avoid naming collisions during CI).  A backfill migration has been added to assign these workspaces to all existing studies.  Additionally, the code used for provisioning & accessing associated workspaces has been substantively refactored to make it more maintainable and reusable.  Some existing validations have been refined (or removed if they were deemed redundant) to help speed up the process of provisioning both workspaces.  Additionally, a migration was added to create the new admin internal group and add all existing admins to this group.  Group membership is managed via a `before_save` hook to add or remove a user if their admin privileges change.

It is worth noting that it takes roughly 35 seconds to provision a Terra workspace.  This means that new study creation can take up to 70 seconds to complete.  Also, given that there are 1600+ studies on production with active Terra workspaces, it will take roughly 2h40m to create all necessary internal workspaces.  Since this is done in the background, it will not affect deployments and should not represent meaningful load on the server while processing.

Parallelization is an option for improving the performance on this, however given the OOO re: accession assignment and other validations, this has the potential for race conditions that could lead to studies being "stuck" or problems in getting new accessions.  Given the relative sparsity of study creation, I opted to [ticket](https://broadworkbench.atlassian.net/browse/SCP-5731) the performance improvements in favor of moving forward with the larger pre-rendering epic.

Lastly, this work only focuses on assigning internal workspaces to new and existing studies.  Forthcoming work will address integrating these workspaces into the normal ingest UX, as well as migrating existing data to the new internal buckets.

#### MANUAL TESTING
1. Boot all services including DelayedJob
2. In a separate terminal, run the backfill migration(s) with `bin/rails db:migrate`
3. Look for log messages like the following in `development.log`
```
FireCloud API request (POST) https://api.firecloud.org/api/workspaces with tracking identifier: default-service-account@broad-singlecellportal-bistlin.iam.gserviceaccount.com
SCP13 Terra internal workspace SCP13-development-internal creation successful
Checking service account permissions for single-cell-portal-development/SCP13-development-internal
FireCloud API request (GET) https://api.firecloud.org/api/groups with tracking identifier: default-service-account@broad-singlecellportal-bistlin.iam.gserviceaccount.com
FireCloud API request (PATCH) https://api.firecloud.org/api/workspaces/single-cell-portal-development/SCP13-development-internal/acl?inviteUsersNotFound=true with tracking identifier: default-service-account@broad-singlecellportal-bistlin.iam.gserviceaccount.com
FireCloud API request (GET) https://api.firecloud.org/api/workspaces/single-cell-portal-development/SCP13-development-internal/acl with tracking identifier: default-service-account@broad-singlecellportal-bistlin.iam.gserviceaccount.com
Setting workspaces acls for single-cell-portal-development/SCP13-development-internal
FireCloud API request (GET) https://api.firecloud.org/api/groups with tracking identifier: default-service-account@broad-singlecellportal-bistlin.iam.gserviceaccount.com
FireCloud API request (PATCH) https://api.firecloud.org/api/workspaces/single-cell-portal-development/SCP13-development-internal/acl?inviteUsersNotFound=true with tracking identifier: default-service-account@broad-singlecellportal-bistlin.iam.gserviceaccount.com
FireCloud API request (PATCH) https://api.firecloud.org/api/workspaces/single-cell-portal-development/SCP13-development-internal/acl?inviteUsersNotFound=true with tracking identifier: default-service-account@broad-singlecellportal-bistlin.iam.gserviceaccount.com
FireCloud API request (PATCH) https://api.firecloud.org/api/workspaces/single-cell-portal-development/SCP13-development-internal/acl?inviteUsersNotFound=true with tracking identifier: default-service-account@broad-singlecellportal-bistlin.iam.gserviceaccount.com
Workspace acls for single-cell-portal-development/SCP13-development-internal configured successfully
SCP13 Terra internal workspace acls assigned successfully
SCP13 Terra internal workspace creation complete
Checking bucket access on single-cell-portal-development/SCP13-development-internal with tracking identifier: default-service-account@broad-singlecellportal-bistlin.iam.gserviceaccount.com
...
```
4. In a browser, sign in and click the "Create study" button in the top navbar
5. Create a new study and confirm that it completes successfully after ~1m and that you are taken to the upload wizard
6. In a separate Rails console session, validate that your admin account is now a member of the new internal admin group
```
user = User.find_by(email: <your admin email>)
group_email = "#{FireCloudClient::ADMIN_INTERNAL_GROUP_NAME}@firecloud.org"
groups = user.user_groups
groups.include?(group_email)
=> true
```
